Mathlib/GroupTheory/Subgroup/MulOpposite.lean

```diff
@@ -17,6 +17,52 @@ subgroup, subgroups
 
 
 variable {G : Type*} [Group G]
+variable {M : Type*} [MulOneClass M]
+
+namespace Submonoid
+
+
+/-- pull a submonoid back to an opposite submonoid along `unop`-/
+@[to_additive (attr := simps)]
+def op {M : Type*} [MulOneClass M] (x : Submonoid M) : Submonoid (MulOpposite M) where
+  carrier := MulOpposite.unop ⁻¹' x.1
+  mul_mem' {a b} ha hb := by
+    simp only [Set.mem_preimage, MulOpposite.unop_mul, SetLike.mem_coe] at ha hb ⊢
+    exact mul_mem hb ha
+  one_mem' := Submonoid.one_mem' _
+
+/-- pull an opposite submonoid back to a submonoid along `op`-/
+@[to_additive (attr := simps)]
+def unop {M : Type*} [MulOneClass M] (x : Submonoid (MulOpposite M)) : Submonoid M where
+  carrier := MulOpposite.op ⁻¹' x.1
+  mul_mem' {a b} ha hb := by
+    simp only [Set.mem_preimage, MulOpposite.unop_mul, SetLike.mem_coe] at ha hb ⊢
+    exact mul_mem hb ha
+  one_mem' := Submonoid.one_mem' _
+
+/-- A submonoid `H` of `G` determines a submonoid `H.opposite` of the opposite group `Gᵐᵒᵖ`. -/
+@[to_additive (attr := simps) "A additive submonoid `H` of `G` determines an additive submonoid
+`H.opposite` of the opposite group `Gᵐᵒᵖ`."]
+def opposite : Submonoid M ≃ Submonoid Mᵐᵒᵖ where
+  toFun := op
+  invFun := unop
+  left_inv _ := SetLike.coe_injective rfl
+  right_inv _ := SetLike.coe_injective rfl
+
+/-- Bijection between a submonoid `H` and its opposite. -/
+@[to_additive (attr := simps!) "Bijection between an additive submonoid `H` and its opposite."]
+def oppositeEquiv (H : Submonoid M) : H ≃ opposite H :=
+  MulOpposite.opEquiv.subtypeEquiv fun _ => Iff.rfl
+
+@[to_additive]
+instance (H : Submonoid M) [Encodable H] : Encodable (opposite H) :=
+  Encodable.ofEquiv H H.oppositeEquiv.symm
+
+@[to_additive]
+instance (H : Submonoid M) [Countable H] : Countable (opposite H) :=
+  Countable.of_equiv H H.oppositeEquiv
+
+end Submonoid
 
 namespace Subgroup
```

Comment:
Can you add the lemma that says `(opposite H).toSubmonoid = opposite H.toSubmonoid` and similar for `opposite.symm`?

Reply:
done